File: bindings/python/examples/governance.py
"""Governance demo.

Starts an in-memory relay, creates an identity, creates a governed context,
executes a role-change governance action, and verifies the result.

Phase 4 PR 5 (#1549) moved governance helpers onto :class:`scp_sdk.SCP`.
Use :meth:`SCP.governance_execute` with a JSON proposal.

Requires a built native extension (``maturin develop --release``).
"""

import asyncio
import json

from scp_sdk import SCP
from scp_sdk.types import Capability, CustodyType, MemoryScope


async def main() -> None:
    with SCP(storage={"type": "in_memory"}) as scp:
        # 1. Start an in-memory relay.
        relay = await scp.relay_start_in_memory()
        print(f"Relay listening at {relay.relay_url}")

        # 1b. Connect transport to the relay.
        await scp.transport_connect(relay.relay_url)

        # 2. Create admin identity.
        admin = await scp.identity_create(CustodyType.IN_MEMORY)
        print(f"Admin DID: {admin.did}")

        # 3. Create a context with governance enabled.
        ctx = await scp.context_create(
            admin.did,
            {
                "ceiling": [
                    Capability.MESSAGES_READ.value,
                    Capability.MESSAGES_WRITE.value,
                    Capability.MEMBER_INVITE.value,
                    Capability.GOVERNANCE_PROPOSE.value,
                    Capability.GOVERNANCE_VOTE.value,
                ],
                "memory_scope": MemoryScope.EPHEMERAL.value,
                "governance": "single_admin",
                "ttl": 600.0,
            },
        )
        print(f"Governed context created: {ctx.context_id}")

        try:
            # 4. Create a second identity and have them join.
            member = await scp.identity_create(CustodyType.IN_MEMORY)
            await scp.context_join(ctx._raw_handle, member.did)
            print(f"Member {member.did} joined")

            # 5. Admin proposes a governance action to change the member's role.
            #    Under SingleAdmin the proposal is tracked, approved, and
            #    executed by the runtime in one step; the response carries the
            #    `proposal_id` the engine retained.
            action = json.dumps(
                {
                    "ChangeRole": {
                        "did": member.did,
                        "new_role": "moderator",
                    },
                }
            )
            propose_result = json.loads(
                await scp.governance_propose(ctx._raw_handle, admin.did, action)
            )
            proposal_id = propose_result["proposal_id"]
            print(f"Proposed + executed governance action: {proposal_id}")

            # `governance_execute` runs an already-approved proposal BY ID. The
            # runtime resolves the authoritative proposal from its own
            # quorum-validated engine — the caller passes no action, only the id.
            # The SingleAdmin propose above already executed this proposal, so a
            # direct execute now demonstrates the runtime's replay guard.
            try:
                await scp.governance_execute(ctx._raw_handle, admin.did, proposal_id)
            except Exception as replay:  # illustrative
                print(f"Re-executing an applied proposal is rejected: {replay}")

            # 6. Cleanup.
            await scp.context_leave(ctx._raw_handle, member.did)
            await scp.context_close(ctx._raw_handle, admin.did)
        finally:
            await relay.shutdown()

    print("Demo complete")


if __name__ == "__main__":
    asyncio.run(main())

File: bindings/typescript/examples/governance.ts
/**
 * Governance demo.
 *
 * Starts an in-memory relay via the caller-owned `SCP`, creates an
 * identity, creates a governed context, executes a governance action
 * (role change), and reports the outcome.
 *
 * Post-Phase-4 (ADR-048): every SDK call routes through an explicit
 * `SCP` instance. The static `Relay.startInMemory()` factory was
 * removed — construct via `scp.relayStartInMemory()`.
 *
 * Run: bun run examples/governance.ts
 */

import { SCP } from "../src/index";

async function main(): Promise<void> {
  const scp = new SCP({ storage: { type: "in_memory" } });
  let relay: Awaited<ReturnType<SCP["relayStartInMemory"]>> | null = null;
  try {
    // 1. Start an in-memory relay.
    relay = await scp.relayStartInMemory();
    console.log(`Relay listening at ${relay.relayUrl}`);

    // 2. Create admin identity.
    const admin = await scp.identityCreate("in_memory");
    console.log(`Admin DID: ${admin.did}`);

    // 3. Wire the bridge's transport to the relay.
    await scp.configureRelayTransport(relay.relayUrl, admin.did);

    // 4. Create a context with governance enabled.
    const ctx = await scp.contextCreate(
      admin,
      JSON.stringify({
        ceiling: [
          "messages:read",
          "messages:write",
          "member:invite",
          "governance:propose",
          "governance:vote",
        ],
        memoryScope: "ephemeral",
        governance: "single_admin",
        ttl: 600,
      }),
    );
    console.log(`Governed context created: ${ctx.contextId}`);

    // 5. Create a second identity and have them join.
    const member = await scp.identityCreate("in_memory");
    await scp.contextJoin(ctx._rawHandle, member.did, null);
    console.log(`Member ${member.did} joined`);

    // 6. Admin proposes a governance action to change the member's role, then
    //    executes the tracked, approved proposal BY ID. Execution takes only
    //    the proposal id — the executor and consequence subject are resolved
    //    from the tracked proposal's proposer, never passed by the caller.
    try {
      const action = JSON.stringify({
        ChangeRole: { did: member.did, new_role: "moderator" },
      });
      const proposeResult = await scp.contextGovernancePropose(
        ctx._rawHandle,
        action,
        admin.did,
      );
      const proposalIdHex = JSON.parse(proposeResult).proposal_id as string;
      console.log(`Proposal created: ${proposalIdHex}`);

      const result = await scp.contextExecuteGovernanceAction(
        ctx._rawHandle,
        proposalIdHex,
      );
      console.log("Governance action result:", result);
    } catch (err) {
      console.log("Governance action:", err);
    }

    // 7. Cleanup.
    await scp.contextClose(ctx._rawHandle, admin.did);
    console.log("Context closed");
  } finally {
    if (relay !== null) {
      await relay.shutdown();
    }
    await scp.shutdown(5);
    console.log("Demo complete");
  }
}

main().catch((error: unknown) => {
  console.error("Demo failed:", error);
  process.exit(1);
});
